datasource/src/main/scala/quasar/physical/s3/impl/evaluate.scala

```diff
@@ -93,17 +103,24 @@ object evaluate {
     // figure out how we're going to parse the object as JSON
     val circeJsonPipe = circePipe[F](jsonParsing)
 
-    sign(request) >>= (r =>
-      streamRequestThroughFs2[F, Data](client, r) { resp =>
-        // convert the data to JSON, using the parsing method
-        // of our choice
-        val asJson: Stream[F, Json] = resp.body.through(circeJsonPipe)
+    sign(request) >>= { r =>
+      val stream =
+        OptionT(streamRequestThroughFs2[F, Data](client, r) { resp =>
+          // convert the data to JSON, using the parsing method
+          // of our choice
+          val asJson: Stream[F, Json] = resp.body.through(circeJsonPipe)
 
-        // convert the JSON from circe's representation to ours
-        val asData: Stream[F, Data] = asJson.map(circeJsonToData)
+          // convert the JSON from circe's representation to ours
+          val asData: Stream[F, Data] = asJson.map(circeJsonToData)
 
-        // and we're done.
-        asData
-      })
+          // and we're done.
+          asData
+        })
+
+      stream.map(_.handleErrorWith {
+        case ParsingFailure(_, _) =>
+          Stream.raiseError(QuasarError.throwableP(noParseError(file)))
```

Comment:
I think we shouldn't lose the exact `ParsingFailure` that happened (message and underlying cause). I.e. this info should be supplied as part of `noParseError` in some way (together with the expected format)).

Reply:
👍

Reply:
I now retain the `message` from `ParsingFailure`. The message from the `underlying` is exactly the same as the message for this particular exception.